HTML media capture sometimes shows a black camera preview for SafariViewServices

youennf opened the pull request:
#### 47387a7ce4cb5cf4ee63f6288c0092dc9855837e
<pre>
HTML media capture sometimes shows a black camera preview for SafariViewServices
<a href="https://bugs.webkit.org/show_bug.cgi?id=292269">https://bugs.webkit.org/show_bug.cgi?id=292269</a>
<a href="https://rdar.apple.com/149678625">rdar://149678625</a>

Reviewed by Eric Carlson.

As part of HTML Media Capture, UIProcess sometimes launches a camera preview directly.
In the case of SafariViewServices, the camera library code will check whether SafariViewServices is in the backgorund or in the foreground.
If it is in foreground, everything is fine. Otherwise, it will mute the camera.

To compute whether SafariViewServices is in the background or not, the camera library code needs to identify which parent application is launching SafariViewServices.
To make sure, the computation is correct, we are now setting the presenting application PID just before launching the camera preview.

We do a small refactoring to pass more references, and constify them if possible.

Manually tested.

* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::handleRunOpenPanel):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::runOpenPanel):
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::handleRunOpenPanel):
* Source/WebKit/UIProcess/win/PageClientImpl.cpp:
(WebKit::PageClientImpl::handleRunOpenPanel):
* Source/WebKit/UIProcess/win/PageClientImpl.h:

Canonical link: <a href="https://commits.webkit.org/294336@main">https://commits.webkit.org/294336@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/65937995ed3347eec0115ef862f481650332b230

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101503 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21169 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11473 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106657 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/52132 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21476 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29663 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77302 "Found 1 new test failure: fast/css/view-transitions-nested-transparency-layers.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34331 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104510 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16575 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91661 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57644 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16398 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51480 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/86282 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9754 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109008 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28632 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21062 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/86272 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28994 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87868 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85839 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21837 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30574 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8282 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22761 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28563 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33844 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28374 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31694 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29933 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
```diff
@@ -741,8 +745,16 @@ - (BOOL)_isHostedInAnotherProcess;
 }
 
-bool PageClientImpl::handleRunOpenPanel(WebPageProxy*, WebFrameProxy*, const FrameInfoData& frameInfo, API::OpenPanelParameters* parameters, WebOpenPanelResultListenerProxy* listener)
+bool PageClientImpl::handleRunOpenPanel(WebPageProxy* proxy, WebFrameProxy*, const FrameInfoData& frameInfo, API::OpenPanelParameters* parameters, WebOpenPanelResultListenerProxy* listener)
```
Nit: it might be good to change the proxy parameter from `WebPageProxy*` to `const WebPageProxy&` now that it is being used.

youennf replied:
Done

pxlcoder commented on Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
```diff
@@ -741,8 +745,16 @@ - (BOOL)_isHostedInAnotherProcess;
 }
 
-bool PageClientImpl::handleRunOpenPanel(WebPageProxy*, WebFrameProxy*, const FrameInfoData& frameInfo, API::OpenPanelParameters* parameters, WebOpenPanelResultListenerProxy* listener)
+bool PageClientImpl::handleRunOpenPanel(WebPageProxy* proxy, WebFrameProxy*, const FrameInfoData& frameInfo, API::OpenPanelParameters* parameters, WebOpenPanelResultListenerProxy* listener)
```
I would name the parameter "page" rather than "proxy":
- That's how `WebPageProxy` parameters are normally named
- There's also a `WebFrameProxy` here

youennf replied:
Done

pxlcoder commented on Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
```diff
@@ -741,8 +745,16 @@ - (BOOL)_isHostedInAnotherProcess;
-bool PageClientImpl::handleRunOpenPanel(WebPageProxy*, WebFrameProxy*, const FrameInfoData& frameInfo, API::OpenPanelParameters* parameters, WebOpenPanelResultListenerProxy* listener)
+bool PageClientImpl::handleRunOpenPanel(WebPageProxy* proxy, WebFrameProxy*, const FrameInfoData& frameInfo, API::OpenPanelParameters* parameters, WebOpenPanelResultListenerProxy* listener)
 {
+    RELEASE_LOG_INFO(WebRTC, "PageClientImpl::handleRunOpenPanel");
```
It's strange to see this in the `WebRTC` category, since this method is more generally used for file/photo upload. But I don't have another suggestion from the existing set of categories.

youennf replied:
Hum, we can change to Media maybe?
I put it as WebRTC since the issue might usually be about camera and microphones not working, and WebRTC encompasses camera and microphone as well.

youennf committed:
HTML media capture sometimes shows a black camera preview for SafariViewServices
https://bugs.webkit.org/show_bug.cgi?id=292269
rdar://149678625

Reviewed by Eric Carlson.

As part of HTML Media Capture, UIProcess sometimes launches a camera preview directly.
In the case of SafariViewServices, the camera library code will check whether SafariViewServices is in the backgorund or in the foreground.
If it is in foreground, everything is fine. Otherwise, it will mute the camera.

To compute whether SafariViewServices is in the background or not, the camera library code needs to identify which parent application is launching SafariViewServices.
To make sure, the computation is correct, we are now setting the presenting application PID just before launching the camera preview.

We do a small refactoring to pass more references, and constify them if possible.

Manually tested.

* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::handleRunOpenPanel):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::runOpenPanel):
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::handleRunOpenPanel):
* Source/WebKit/UIProcess/win/PageClientImpl.cpp:
(WebKit::PageClientImpl::handleRunOpenPanel):
* Source/WebKit/UIProcess/win/PageClientImpl.h:

Canonical link: https://commits.webkit.org/294336@main
